Add libxmp_crc32c function and fast module hashing mode.

AliceLR opened the pull request:
Adds a CRC-32C calculation function simplified from Mark Adler's fast CRC-32C function, with fast hardware ARM and RISC-V Zbc implementations added.

Also adds an optional fast module hashing mode to libxmp, allowing the MD5 to be skipped unless a module matches both the expected length and CRC-32C of a module in the quirks list. This reduces loading time significantly for most modules, at the cost of users opting out of being able to use the MD5 for their own module quirks lists. Fast hashing mode is intended to accelerate fuzzing and loading on embedded platforms, and is disabled by default.

Finally, I've removed an old copy of nebulos by Audiomonster from the quirks list. I've been unable to locate this copy, and all existing copies are correctly detected as VBlank by scan compare.

See libxmp issue #400 for more information.

TODO:
* Locate missing copy of "No Mercy" by Alf.
* Add xmp_set_player setting to configure hashing mode and update documentation.
* Win32 detection of ARMv8.0 extensions?
* Define `LIBXMP_NO_HARDWARE_CRC` in Makefile.vc and watcom.mif?
* CMake detection of `getauxval`/`asm/hwcap.h`/`sysctlbyname`.
* Vectorized x86 CRC-32C implementation. Newer versions of Mark Adler's CRC-32C use the faster vectorized method so this shouldn't be much trouble to adapt. This increases the architecture requirements of the algorithm slightly (Westmere rather than Nehalem).
* Vectorized ARM CRC-32C implementation. This needs an extra `getauxval`/`sysctlbyname` call, as `PMULL` was *also* optional in ARMv8.0.
* Fix unfinished RISC-V Zbc CRC-32C implementation. I'm not aware of any useful boards that implement B yet, and Linux hasn't added a HWCAP flag for B or Zbc yet, so this is mostly just for QEMU :^)

When the fast routine is enabled, I've seen this reduce full loads of the Modland Protracker directory from ~50s to ~30s (i7-7700, Windows 10, test-dev/xmpchk, filesystem cached in RAM). Fuzzing improvement won't be quite this good since most inputs already fail to load, and the fuzzing routine now uses the CRC-32C to seed player settings. All fast hashing code is currently disabled in the patch to pass the regression tests.

sezero commented:
Some notes:

- Build failure on i686-linux:
```
src/crc32c.c: In function ‘libxmp_crc32c’:
src/crc32c.c:331: error: PIC register ‘%ebx’ clobbered in ‘asm’
```

- ix86 in general: detect `cpuid` existence first?

- OW, x86: has its own inline asm syntax (`#pragma aux ....`)

- MSVC, x86: has its own inline asm syntax

- MSVC, x64: has no inline asm at all. uses intrinsics instead
  as of VS 2008 version. (include intrin.h or possibly nmmintrin.h for
  `_mm_crc32_u8|16|32` )

AliceLR commented:
> Some notes:

Thanks, I will check into this. I don't know if I'll get back to this right away. This branch is a tumor that grew off of MOD VBlank detection and my old CRC-32C testing, I had to get it out of my working tree :-)

sezero commented:
For calling cpuid, and also the PIC issue, see SDL's code:
https://github.com/libsdl-org/SDL/blob/SDL2/src/cpuinfo/SDL_cpuinfo.c

sezero commented:
For RISC stuff, CC: @ccawley2011

ccawley2011 commented:
> For RISC stuff, CC: @ccawley2011

I don't know anything about RISC-V. I'm more familiar with RISC OS, which is unrelated.

That said, it would be interesting to get the ARMv8 code working on RISC OS, although GCC 4.7.4 predates the `arm_acle.h` header so it may be necessary to provide an inline ASM fallback for that particular use case.

AliceLR commented:
> * OW, x86: has its own inline asm syntax (`#pragma aux ....`)
> * MSVC, x86: has its own inline asm syntax
> * MSVC, x64: has no inline asm at all. uses intrinsics instead
> GCC 4.7.4 predates the `arm_acle.h` header so it may be necessary to provide an inline ASM fallback for that particular use case.

I suspect supporting all of these is going to get messy—maybe an `src/crc32c/` and a header per-implementation is justified...

> That said, it would be interesting to get the ARMv8 code working on RISC OS

Does RISC OS allow detection of the required CPU features at runtime? (System call, user code runs at EL1 or higher, RISC OS guarantees access to `ID_AA64ISAR0_EL1` at EL0, RISC OS allows EL0 to set a custom fault handler, etc?). Due to the current situation with RISC-V I've been considering a compile option to remove the runtime feature check. I don't know why ARM and RISC-V both decided to make this difficult.

ccawley2011 commented:
> Does RISC OS allow detection of the required CPU features at runtime? (System call, user code runs at EL1 or higher, RISC OS guarantees access to `ID_AA64ISAR0_EL1` at EL0, RISC OS allows EL0 to set a custom fault handler, etc?).

It should be possible using the [OS_PlatformFeatures](https://www.riscosopen.org/wiki/documentation/show/OS_PlatformFeatures) SWI.

AliceLR commented:
I fixed the RISC-V CRC-32C implementation and fixed minor issues that caused some of the Linux regression tests to fail. I have a much better idea of how to handle the vectorized x86 implementation now, and I'll fix the MSVC/Watcom tests when I get to that.

I found this regarding an [ARM vectorized implementation](https://developer.arm.com/documentation/dui0801/g/A64-SIMD-Vector-Instructions/PMULL--PMULL2--vector-) though:
> The PMULL instruction extracts each source vector from the lower half of each source register, while the PMULL2 instruction extracts each source vector from the upper half of each source register.
>
> Depending on the settings in the CPACR_EL1, CPTR_EL2, and CPTR_EL3 registers, and the current Security state and Exception level, an attempt to execute the instruction might be trapped.

`PMULL` and `PMULL2` can be detected by the same `getauxval` call as the CRC-32 instructions, but I don't know if these access control registers are going to be a problem.
